File: 941-SortArrayByParity/941-SortArrayByParity.js
// Last updated: 3/18/2026, 11:03:51 PM
/**
 * @param {number[]} nums
 * @return {number[]}
 */

 // brute force 

//     var sortArrayByParity = function(nums) {
//     let result = [];
    
//     // First push even numbers
//     for (let i = 0; i < nums.length; i++) {
//         if (nums[i] % 2 === 0) {
//             result.push(nums[i]);
//         }
//     }
    
//     // Then push odd numbers
//     for (let i = 0; i < nums.length; i++) {
//         if (nums[i] % 2 !== 0) {
//             result.push(nums[i]);
//         }
//     }
    
//     return result;
// };


 // optimised methods

var sortArrayByParity = function(nums) {
    let left = 0, right = nums.length - 1;
    
    while (left < right) {
        if (nums[left] % 2 > nums[right] % 2) {
            // Swap if left is odd and right is even
            [nums[left], nums[right]] = [nums[right], nums[left]];
        }
        
        if (nums[left] % 2 === 0) left++;
        if (nums[right] % 2 === 1) right--;
    }
    
    return nums;
};

File: https:/github.com/Raushankumar0720/Leetcode/941-SortArrayByParity/941-SortArrayByParity.cpp
// Last updated: 8/19/2026, 10:00:32 PM
class Solution {
public:
    vector<int> sortArrayByParity(vector<int>& nums) {

        vector<int> even;
        vector<int> odd;

        for (int num : nums) {
            if (num % 2 == 0) {
                even.push_back(num);
            } else {
                odd.push_back(num);
            }
        }

        for (int num : odd) {
            even.push_back(num);
        }

        return even;
    }
};